spec/forms/challenge_partnership_form_spec.rb

```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+RSpec.describe ChallengePartnershipForm, type: :model do
+  describe "validations" do
+    it do
+      is_expected.to validate_presence_of(:challenge_reason)
+        .with_message("Select a reason why you think this confirmation is incorrect")
+    end
+  end
+
+  describe "#challenge!" do
+    let(:lead_provider) { create :lead_provider }
+    let!(:lead_provider_profiles) { create_list(:lead_provider_profile, rand(2..3), lead_provider: lead_provider) }
+    let(:partnership) { create :partnership, lead_provider: lead_provider }
+    let(:reason) { described_class.new.challenge_reason_options.sample.id }
+
+    subject { described_class.new(partnership: partnership, challenge_reason: reason) }
+
+    it "marks given partnership as challenged" do
+      expect { subject.challenge! }.to change { partnership.reload.challenged? }.to true
+    end
+
+    it "schedules partnership challenged emails" do
+      subject.challenge!
+
+      lead_provider_profiles.each do |lp_profile|
+        expect(LeadProviderMailer).to delay_email_delivery_of(:partnership_challenged_email)
+          .with(user: lp_profile.user, partnership: partnership)
+      end
```

Comment:
We could also stub the mailer class like we do in sessions spec and check the email method is being called with the correct arguments. 

I guess what this does extra is it checks that we didn't forget to call `deliver` on the email, but maybe we could also check that by using `TestMailer`?

You know my opinion on complicated matchers that stem from delayed job :P

Reply:
There is one extra check that we do here - we actually check that the email is scheduled for delivery. Any stubbing/mocking can only check that the method `deliver` was called - this is not enough to be certain that the mail is scheduled. If the call is executed within a database transaction, then the job will not be committed and we have false positive. This matcher checks that the appropriate job has been committed to the database.

Reply:
Right, and what is the non-engineered use case when that false positive happens? A dev adds some code after scheduling the email, and that code breaks the transaction, so the email is not scheduled but we think in this test that it is?

Reply:
Commons scenario is that we schedule emails together with some database changes. Example would be partnership creation - we save a number of partnership in the database and schedule an email for each one of them. It was requested that the process is all or none - so we have to wrap the entire loop in the transaction. There was no way of testing that no emails will be sent without going directly to the databse.

What's also worth noting is that this code is much simpler than what ActiveJob is doing behind the curtain. This code is highly afferent and as such is "write it once, improve twice and never see it again".

Reply:
What the new matcher is giving us is the ability to check that an email was delayed successfully with particular parameters, and it will not succeed when transaction breaks. 

But really, should we not have a separate test for the transaction breaking (and then we can check there are no new delayed jobs), and another test for the happy path, in which we can check that the mailer got called with the correct parameters? And even check that exactly one new job got scheduled?

I know you know it, but I am weary of delving into the depths of delayed job and how it stores things in the database. And to your point that this code is essentially forgettable - we hoped so about `delay_execution_of`, but it is actually [touched](https://github.com/DFE-Digital/early-careers-framework/pull/311/files#diff-dc4a9b45ec5529e4af2ff7db9553384327f1079f3765e96eaebd21a8f51d55b4R61-R67), cause we couldn't predict everything :/

Reply:
`But really, should we not have a separate test for the transaction breaking (and then we can check there are no new delayed jobs), and another test for the happy path, in which we can check that the mailer got called with the correct parameters? And even check that exactly one new job got scheduled?
`

That's exactly the point of this. If we use ActiveJob test queue adapter, no job would be ever written to DJ, regardless of the success. Naturally those are the new matchers and they have not been written to satisfy every single scenario yet - that's where "improve twice" comes from. It will eventually reach its stable form after which no further changes are needed.

But yes, absolutely agree it would be best implemented within the ActiveJob itself. I've really given up on suggesting new features/improvements to rails team and being shot down with "we don't need it in bootscamp" argument...